examples/websocket_async_echo_server.hpp

```diff
@@ -0,0 +1,375 @@
+            boost::asio::io_service::strand strand;
+            beast::websocket::opcode op;
+            beast::streambuf db;
+            std::size_t id;
+
+            data(async_echo_server& server_,
+                endpoint_type const& ep_,
+                    socket_type&& sock_)
+                : server(server_)
+                , ep(ep_)
+                , ws(std::move(sock_))
+                , strand(ws.get_io_service())
+                , id([]
+                    {
+                        static std::atomic<std::size_t> n{0};
+                        return ++n;
+                    }())
+            {
+            }
+        };
+
+        // VFALCO This could be unique_ptr in [Net.TS]
+        std::shared_ptr<data> d_;
+
+    public:
+        peer(peer&&) = default;
+        peer(peer const&) = default;
+        peer& operator=(peer&&) = delete;
+        peer& operator=(peer const&) = delete;
+
+        template<class... Args>
+        explicit
+        peer(async_echo_server& server,
+            endpoint_type const& ep, socket_type&& sock,
+                Args&&... args)
+            : d_(std::make_shared<data>(server, ep,
+                std::forward<socket_type>(sock),
+                    std::forward<Args>(args)...))
+        {
+            auto& d = *d_;
+            d.server.opts_.set_options(d.ws);
+            run();
+        }
+
+        void run()
+        {
+            auto& d = *d_;
+            d.ws.async_accept(std::move(*this));
+        }
+
+        void operator()(error_code ec, std::size_t)
+        {
+            (*this)(ec);
+        }
+
+        void operator()(error_code ec)
+        {
+            using boost::asio::buffer;
+            using boost::asio::buffer_copy;
+            auto& d = *d_;
+            switch(d.state)
+            {
+            // did accept
+            case 0:
+                if(ec)
+                    return fail("async_accept", ec);
+
+            // start
+            case 1:
+                if(ec)
+                    return fail("async_handshake", ec);
+                d.db.consume(d.db.size());
+                // read message
+                d.state = 2;
+                d.ws.async_read(d.op, d.db,
+                    d.strand.wrap(std::move(*this)));
+                return;
+
+            // got message
+            case 2:
+                if(ec == beast::websocket::error::closed)
+                    return;
+                if(ec)
+                    return fail("async_read", ec);
+                // write message
+                d.state = 1;
+                d.ws.set_option(
+                    beast::websocket::message_type(d.op));
+                d.ws.async_write(d.db.data(),
+                    d.strand.wrap(std::move(*this)));
+                return;
+            }
+        }
+
+    private:
+        void
+        fail(std::string what, error_code ec)
+        {
+            auto& d = *d_;
+            if(d.server.log_)
```

Comment:
does it make sense for this method the method in the server class to both being checking `log_` ?